kong/plugins/hmac-auth/access.lua

```diff
@@ -157,11 +157,48 @@ local function create_hash(request, hmac_params, headers)
   return hmac[hmac_params.algorithm](hmac_params.secret, signing_string)
 end
 
+-- plugin assumes the request parameters being used for creating
+-- signature by client are not changed by core or any other plugin
+-- DEPRECATED BY: https://github.com/Kong/kong/pull/3339
+local function create_hash_deprecated(request, hmac_params, headers)
+  local signing_string = ""
+  local hmac_headers = hmac_params.hmac_headers
+  local count = #hmac_headers
+
+  for i = 1, count do
+    local header = hmac_headers[i]
+    local header_value = headers[header]
+
+    if not header_value then
+      if header == "request-line" then
+        -- request-line in hmac headers list
+        local request_line = fmt("%s %s HTTP/%s", ngx.req.get_method(), ngx.var.uri, ngx.req.http_version())
+
+        signing_string = signing_string .. request_line
+      else
+        signing_string = signing_string .. header .. ":"
+      end
+    else
+      signing_string = signing_string .. header .. ":" .. " " .. header_value
+    end
+    if i < count then
+      signing_string = signing_string .. "\n"
+    end
+  end
+  return hmac[hmac_params.algorithm](hmac_params.secret, signing_string)
+end
+
 local function validate_signature(request, hmac_params, headers)
-  local signature_1 = create_hash(request, hmac_params, headers)
-  local signature_2 = ngx_decode_base64(hmac_params.signature)
+  local signature_1            = create_hash(request, hmac_params, headers)
+  local signature_2            = ngx_decode_base64(hmac_params.signature)
```

Comment:
ditto

Reply:
Also we don't need to create both the signature at once, one should be fall back and only be created if needed.

Reply:
@shashiranjan84 Whitespace removed, My commit from yesterday should have included the deprecated signature creation into the conditional. This looks like it would do what you're describing.

Reply:
You are right. Thanks again for your effort, we just need add test, which would be very similar to https://github.com/Kong/kong/blob/master/spec/03-plugins/20-hmac-auth/03-access_spec.lua#L1150, you would just need to create request signature without query-param

Reply:
Test Added